pkg/core/chain/chain.go

```diff
@@ -411,24 +407,83 @@ func (c *Chain) startConsensus() error {
 // VerifyCandidateBlock can be used as a callback for the consensus in order to
 // verify potential winning candidates.
 func (c *Chain) VerifyCandidateBlock(hash []byte) error {
-	var cm *message.Candidate
-	if err := c.db.View(func(t database.Transaction) error {
-		var err error
-		cm, err = t.FetchCandidateMessage(hash)
-		return err
-	}); err != nil {
+	cm, err := c.fetchCandidateMessage(hash)
+	if err != nil {
 		return err
 	}
 
 	// We first perform a quick check on the Block Header and
-	if err := c.verifier.SanityCheckBlock(*c.tip, *cm.Block); err != nil {
+	if err = c.verifier.SanityCheckBlock(*c.tip, *cm.Block); err != nil {
 		return err
 	}
 
-	_, err := c.proxy.Executor().VerifyStateTransition(c.ctx, cm.Block.Txs, cm.Block.Header.Height)
+	_, err = c.proxy.Executor().VerifyStateTransition(c.ctx, cm.Block.Txs, cm.Block.Header.Height)
 	return err
 }
 
+func (c *Chain) fetchCandidateMessage(hash []byte) (message.Candidate, error) {
+	var cm message.Candidate
+	err := c.db.View(func(t database.Transaction) error {
+		var err error
+		cm, err = t.FetchCandidateMessage(hash)
+		return err
+	})
+
+	if err != nil && err != database.ErrBlockNotFound {
+		return message.Candidate{}, err
+	}
+
+	// If the candidate message isn't found, we will ask our direct peers for it.
+	if err == database.ErrBlockNotFound {
+		cm, err = c.requestCandidate(hash)
+		if err != nil {
+			return message.Candidate{}, err
+		}
+
+		// Store the candidate for future use
+		err = c.db.Update(func(t database.Transaction) error {
+			return t.StoreCandidateMessage(cm)
+		})
+	}
+
+	return cm, err
+}
+
+func (c *Chain) requestCandidate(hash []byte) (message.Candidate, error) {
+	// Make sure we get temporarily notified of incoming messages
+	candidateChan := make(chan message.Message, 10)
+	cChan := eventbus.NewChanListener(candidateChan)
+	id := c.eventBus.Subscribe(topics.Candidate, cChan)
+	defer c.eventBus.Unsubscribe(topics.Candidate, id)
```

Comment:
This is pretty ugly and signals that `EventBus` subscription and asynchronous notification in general is not the answer. What you might want to consider here is a RPCBus call with a timeout (maybe managed through a context cancelation) to a different component that handles the peer requests to the candidates and is completely transparent to the `Chain`. Such component would deal with the `Candidate` topic. 

Moreover, I am unsure why this logic is in the Chain. I would argue that maybe we should keep the logic related to candidate fetching in one place, rather than scattering it around. The most suitable place might be the consensus, since it must deal with the block candidate verification. To that respect, we could change the `loop.Spin` function to return the certificates and the candidate block. This way, the Chain would not be concerned with that.